source/hpack/decoder.d

```diff
@@ -0,0 +1,278 @@
+		else m_range = range;
+		m_table = table;
+
+		decode();
+	}
+
+// InputRange specific methods
+	@property bool empty() @safe { return m_decoded.empty; }
+
+	@property HTTP2HeaderTableField front() @safe { return m_decoded.front; }
+
+	@property HTTP2HeaderTableField back() @safe { return m_decoded.back; }
+
+	void popFront() @trusted
+	{
+		enforce!HPACKException(!empty, "Cannot call popFront on an empty HeaderDecoder");
+
+		m_decoded.popFront();
+
+		// advance if data is still available
+		if(!m_range.empty) decode();
+	}
+
+	@property HTTP2HeaderTableField opIndex(int idx) @safe
+	{
+		assert(idx < m_decoded.length, "Invalid decoder index");
+		return m_decoded[idx];
+	}
+
+	@property HTTP2HeaderTableField[] opSlice(int start, int end) @safe
+	{
+		assert(start >= 0 && end < m_decoded.length, "Invalid decoder slice");
+		return m_decoded[start..end];
+	}
+
+	void put(T)(T range) @safe
+	{
+		static if(is(ElementType!T : char)) m_range ~= cast(ubyte[])range;
+		else m_range ~= range;
+
+		decode();
+	}
+
+	@property HTTP2HeaderTableField[] toIndex() @safe { return m_index; }
+
+	@property HTTP2HeaderTableField[] neverIndexed() @safe { return m_noindex; }
+
+// decoding
+	private {
+
+		void decode() @trusted
+		{
+			while(!m_range.empty) {
+
+				auto bbuf = [m_range[0]].toBitArray();
+				m_range = m_range[1..$];
+
+
+				switch(bbuf[0]) {
+					case 1: // indexed (integer)
+						auto res = decodeInteger(bbuf);
+						m_decoded ~= m_table[res];
+						break;
+					case 0: // literal
+						HTTP2HeaderTableField hres;
+						bool update = false;
+
+						if (bbuf[1]) { // inserted in dynamic table
+							auto idx = bbuf.toInteger(2);
+							if(idx > 0) {  // name == table[index].name, value == literal
+								hres.name = m_table[idx].name;
+							} else {   // name == literal, value == literal
+								hres.name = decodeLiteral();
+							}
+							hres.value = decodeLiteral();
+							m_index ~= hres;
+
+						} else if(bbuf[3]) { // NEVER inserted in dynamic table
+							auto idx = bbuf.toInteger(4);
+							if(idx > 0) {  // name == table[index].name, value == literal
+								hres.name = m_table[idx].name;
+							} else {   // name == literal, value == literal
+								hres.name = decodeLiteral();
+							}
+							hres.value = decodeLiteral();
+							m_noindex ~= hres;
+
+						} else if(!bbuf[2]) { // this occourrence is not inserted in dynamic table
+							auto idx = bbuf.toInteger(4);
+							if(idx > 0) {  // name == table[index].name, value == literal
+								hres.name = m_table[idx].name;
+							} else {   // name == literal, value == literal
+								hres.name = decodeLiteral();
+							}
+							hres.value = decodeLiteral();
+
+						} else { // dynamic table size update (bbuf[2] is set)
+							update = true;
+							auto nsize = bbuf.toInteger(3);
+							m_table.updateSize(cast(HTTP2SettingValue)nsize);
+						}
+
+						if(!update) m_decoded ~= hres;
+						break;
+
+					default:
```

Comment:
Isn't `typeof(bbuf[0])` `bool` anyway, so that a simple `if (bbuf[0]) {} else {}` would be sufficient here?